crates/external/rpc/src/traits.rs

```diff
@@ -51,16 +72,117 @@ pub trait ExecutingMessageValidator {
+        let fut = async { supervisor.check_messages(messages.to_vec(), safety).await };
 
         // Await the validation future with timeout.
-        tokio::time::timeout(timeout, fut)
+        match tokio::time::timeout(timeout, fut)
             .await
             .map_err(ExecutingMessageValidatorError::ValidationTimeout)?
+        {
+            Ok(_) => Ok(()),
+            Err(err) => match err {
+                ClientError::Call(err) => Err(err.into()),
+                _ => Err(ExecutingMessageValidatorError::RpcClientError(err)),
+            },
+        }
+    }
+}
+
+impl From<ErrorObjectOwned> for ExecutingMessageValidatorError {
+    fn from(value: ErrorObjectOwned) -> Self {
+        // todo: match on error code instead of message string once resolved <https://github.com/ethereum-optimism/optimism/issues/14603>
+        // Check if it's invalid message call, message example:
+        // `failed to check message: failed to check log: unknown chain: 14417`
+        if value.message().contains(UNKNOWN_CHAIN_MSG) {
+            if let Ok(chain_id) =
+                value.message().split(' ').last().expect("message contains chain id").parse::<u64>()
+            {
+                return Self::UnknownChain(chain_id);
+            }
+        // Check if it's `does not meet the minimum safety` error, message example:
+        // `message {0x4200000000000000000000000000000000000023 4 1 1728507701 901}
+        // (safety level: unsafe) does not meet the minimum safety cross-unsafe"`
+        } else if value.message().contains(MINIMUM_SAFETY_MSG) {
+            let message_safety = if value.message().contains("safety level: safe") {
+                SafetyLevel::Safe
+            } else if value.message().contains("safety level: local-safe") {
+                SafetyLevel::LocalSafe
+            } else if value.message().contains("safety level: cross-unsafe") {
+                SafetyLevel::CrossUnsafe
+            } else if value.message().contains("safety level: unsafe") {
+                SafetyLevel::Unsafe
+            } else if value.message().contains("safety level: invalid") {
+                SafetyLevel::Invalid
+            } else {
+                // Unexpected level name, return generic error
+                return Self::SupervisorServerError(value);
+            };
+            let expected_safety = if value.message().contains("safety finalized") {
+                SafetyLevel::Finalized
+            } else if value.message().contains("safety safe") {
+                SafetyLevel::Safe
+            } else if value.message().contains("safety local-safe") {
+                SafetyLevel::LocalSafe
+            } else if value.message().contains("safety cross-unsafe") {
+                SafetyLevel::CrossUnsafe
+            } else if value.message().contains("safety unsafe") {
+                SafetyLevel::Unsafe
+            } else if value.message().contains("safety invalid") {
+                SafetyLevel::Invalid
```

Comment:
this branch is redundant since all messages will be at least of invalid safety level